prow/hook/server_test.go

```diff
@@ -334,3 +386,180 @@ func TestNeedDemux(t *testing.T) {
+		"kubernetes/test-infra": {
+			{
+				Name:     "coffee",
+				Endpoint: "/coffee",
+			},
+		},
+		"kubernetes/kubernetes": {
+			{
+				Name:     "gumbo",
+				Endpoint: "/gumbo",
+				Events:   []string{"issue_comment"},
+			},
+		},
+		"kubernetes": {
+			{
+				Name:     "chicken",
+				Endpoint: "/chicken",
+				Events:   []string{"repository"},
+			},
+			{
+				Name:     "water",
+				Endpoint: "/water",
+			},
+			{
+				Name:     "chocolate",
+				Endpoint: "/chocolate",
+				Events:   []string{"pull_request", "issue_comment", "repository"},
+			},
+		},
+	}
+
+	// This is the SHA1 signature for payload "$BODY" and signature "abc"
+	// echo -n $BODY | openssl dgst -sha1 -hmac abc
+	const hmac string = "sha1=d5f926df2d39006bdb5b6acb18f8fcdebad7a052"
+	const body string = `{
+  "action": "edited",
+  "changes": {
+    "default_branch": {
+      "from": "master"
+    }
+  },
+  "repository": {
+    "full_name": "kubernetes/test-infra",
+    "default_branch": "master"
+  }
+}`
+
+	metrics := githubeventserver.NewMetrics()
+	pa := &plugins.ConfigAgent{}
+	pa.Set(&plugins.Configuration{
+		ExternalPlugins: externalPlugins,
+	})
+
+	var testcases = []struct {
+		name string
+
+		Method string
+		Header map[string]string
+		Body   string
+
+		ExpectedDispatch []string
+	}{
+		{
+			name: "Repository Event",
+
+			Method: http.MethodPost,
+			Header: map[string]string{
+				"X-GitHub-Event":    "repository",
+				"X-GitHub-Delivery": "I am unique",
+				"X-Hub-Signature":   hmac,
+				"content-type":      "application/json",
+			},
+			Body: body,
+
+			ExpectedDispatch: []string{"/coffee", "/water", "/chicken", "/chocolate"},
+		},
+		{
+			name: "Issue comment Event",
+
+			Method: http.MethodPost,
+			Header: map[string]string{
+				"X-GitHub-Event":    "issue_comment",
+				"X-GitHub-Delivery": "I am unique",
+				"X-Hub-Signature":   hmac,
+				"content-type":      "application/json",
+			},
+			Body: body,
+
+			ExpectedDispatch: []string{"/coffee", "/water", "/chocolate"},
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Logf("Running scenario %q", tc.name)
+
+		var calledExternalPlugins []string
+		client := newTestClient(func(req *http.Request) *http.Response {
+			calledExternalPlugins = append(calledExternalPlugins, req.URL.String())
+			return &http.Response{
+				StatusCode: 200,
+				Body:       ioutil.NopCloser(bytes.NewBufferString(`OK`)),
+				Header:     make(http.Header),
+			}
+		})
+
+		s := &Server{
+			Metrics:        metrics,
+			Plugins:        pa,
+			TokenGenerator: getSecret,
+			RepoEnabled:    func(org, repo string) bool { return true },
+			c:              *client,
+		}
+		w := httptest.NewRecorder()
+		r, err := http.NewRequest(tc.Method, "", strings.NewReader(tc.Body))
+		if err != nil {
+			t.Fatal(err)
+		}
+		for k, v := range tc.Header {
+			r.Header.Set(k, v)
+		}
+		s.ServeHTTP(w, r)
+		s.wg.Wait()
+
+		if !containSameElements(tc.ExpectedDispatch, calledExternalPlugins) {
```

Comment:
Again for legibility in the error case, please use `cmp`:
```suggestion
                if diff := cmp.Diff(tc.ExpectedDispatch, calledExternalPlugin, cmpopts.SortSlices(a, b string) bool { return a < b })); diff != "" {
		if !containSameElements(tc.ExpectedDispatch, calledExternalPlugins) {
```

Reply:
Thank you for the suggestion. I didn't know about `cmp`. Done in 99a292f9e5